velox/dwio/dwrf/reader/SelectiveTimestampColumnReader.cpp

```diff
@@ -146,7 +147,13 @@ void SelectiveTimestampColumnReader::readHelper(
           nanos *= 10;
         }
       }
-      auto seconds = secondsData[i] + EPOCH_OFFSET;
+      int64_t seconds = 0;
+      if (sessionTimezone_) {
```

Comment:
@Yuhta I don't recommend this implementation here. I would like to refer to Presto and introduce the `hive.time-zone` parameter. In order to be compatible with meta internal use cases, the default value is set to `America/Los_Angeles`. Do you have any suggestions?

Reply:
`time_zone::to_sys` has some cost and we don't know the performance impact on the overall workload yet, so better keep the original behavior as much as possible

Reply:
I changed `rawTs[i].toGMT(*sessionTimezone_);` into the following code:
```
auto timezoneOffset =
Timestamp::getTimezoneOffset(rawTs[i], sessionTimezone_);
rawTs[i] = Timestamp(seconds - timezoneOffset, nanos);
```

Reply:
Actually I don't think we should do `toGMT` or `getTimezoneOffset` here and during write, the values in `Timestamp` is always UTC no matter what timezone session property is set.  The only hack is for the Meta internal which use a wrong epoch, but other than that the reader and writer should just use UTC.

Reply:
Do you mean to modify `TimestampWriter` and `TimestampReader` to the following?
```
FOLLY_ALWAYS_INLINE int64_t formatTime(int64_t seconds, uint64_t nanos) {
  DWIO_ENSURE(seconds >= MIN_SECONDS);

 ....
  if (!sessionTimezone) {
     // Compatible with meta internal use cases.
     return seconds - EPOCH_OFFSET;
  }
  
  return seconds;
}
```
```
void fillTimestamps(
    const uint64_t* nullsPtr,
    const int64_t* secondsPtr,
    const uint64_t* nanosPtr,
    vector_size_t numValues) {
    vector_size_t numValues,
    const date::time_zone* sessionTimezone) {
   ...
  for (vector_size_t i = 0; i < numValues; i++) {
      ....

      int64_t seconds;
      if (!sessionTimezone) {
        // Compatible with meta internal use cases.
        seconds = secondsPtr[i] + dwio::common::EPOCH_OFFSET;
      }
      if (seconds < 0 && nanos != 0) {
        seconds -= 1;
      }
      timestamps[i] = Timestamp(seconds, nanos);
    }
  }
}
```

Reply:
I checked the read and write logic of Presto ORC Timestamp. The EPOCH_OFFSET used by Timestamp will takes sessionTimezone into account. https://github.com/prestodb/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java#L135

https://github.com/prestodb/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java#L107

Reply:
In Presto Java, timestamp `Block` is storing values from local epoch, but in Velox, `Timestamp` vectors store UTC epoch, so the reader and writer implementations should be different.

Reply:
I have the following thoughts on the time zone issue of DWRF/ORC reading and writing Timestamp. Can you help me see if it is correct?
1. If `legacy_timestamp`(presto)/`adjust_timestamp_to_session_timezone`(presto c++) is true, and `sessionTimezone` is not empty, then reading and writing Timestamp requires adding sessionTimezone information;
2. If `legacy_timestamp`(presto)/`adjust_timestamp_to_session_timezone`(presto c++) is false, and `sessionTimezone` is not empty, just read and write Timestamp directly;
3. If `legacy_timestamp`(presto)/`adjust_timestamp_to_session_timezone`(presto c++) is false, and `sessionTimezone` is empty, add the offset of `America/Los_Angeles` time zone to be compatible with the business inside meta.
The code corresponding to the above logic is as follows:

READ:
```
template <bool isDense>
void SelectiveTimestampColumnReader::readHelper(
    common::Filter* filter,
    RowSet rows) {
  .....
  auto rawTs = tsValues->asMutable<Timestamp>();

  for (vector_size_t i = 0; i < numValues_; i++) {
      ....
      int64_t seconds = 0;
      if (sessionTimezone_) {
        seconds = secondsData[i] + UTC_EPOCH_OFFSET;
      } else {
        // Compatible with meta internal use cases.
        seconds = secondsData[i] + EPOCH_OFFSET;
      }
      
      ... 

      rawTs[i] = Timestamp(seconds, nanos);
      if (adjustTimestampToTimezone && sessionTimezone_) {
        auto timezoneOffset =
            Timestamp::getTimezoneOffset(rawTs[i], sessionTimezone_);
        rawTs[i] = Timestamp(seconds - timezoneOffset, nanos);
      }
    }
  }
}
```

WRITE:
```
FOLLY_ALWAYS_INLINE int64_t
formatTime(const Timestamp& timestamp, const date::time_zone* sessionTimezone) {
  ....
  if (adjustTimestampToTimezone && sessionTimezone) {
    return seconds - UTC_EPOCH_OFFSET +
        Timestamp::getTimezoneOffset(timestamp, sessionTimezone);
  } else if (!adjustTimestampToTimezone && sessionTimezone) {
    return seconds - UTC_EPOCH_OFFSET;
  }
  // Compatible with meta internal use cases.
  return seconds - EPOCH_OFFSET;
}
```

Reply:
If the `sessionTimezone` can be set to `America/Los_Angeles` in meta, it can be further simplified:
READ:
```
template <bool isDense>
void SelectiveTimestampColumnReader::readHelper(
    common::Filter* filter,
    RowSet rows) {
  .....
  auto rawTs = tsValues->asMutable<Timestamp>();

  for (vector_size_t i = 0; i < numValues_; i++) {
      ....
      int64_t seconds = secondsData[i] + UTC_EPOCH_OFFSET;
      
      ... 

      rawTs[i] = Timestamp(seconds, nanos);
      if (adjustTimestampToTimezone && sessionTimezone_) {
        auto timezoneOffset =
            Timestamp::getTimezoneOffset(rawTs[i], sessionTimezone_);
        rawTs[i] = Timestamp(seconds - timezoneOffset, nanos);
      }
    }
  }
}
```

WRITE:
```
FOLLY_ALWAYS_INLINE int64_t
formatTime(const Timestamp& timestamp, const date::time_zone* sessionTimezone) {
  ....
  if (adjustTimestampToTimezone && sessionTimezone) {
    return seconds - UTC_EPOCH_OFFSET +
        Timestamp::getTimezoneOffset(timestamp, sessionTimezone);
  }
  return seconds - UTC_EPOCH_OFFSET;
}
```

Reply:
I think that is good.  Just instead of using `getTimezoneOffset`, you should use the existing `Timestamp:toGMT` and `Timestamp::toTimezone`